rio_stac/stac.py

```diff
@@ -53,27 +56,105 @@ def get_metadata(
 
     metadata["bbox"] = list(bbox)
     metadata["footprint"] = bbox_to_geom(bbox)
+    return metadata
+
+
+def get_projection_info(
+    src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT, MemoryFile]
+) -> Dict:
+    """Get projection metadata.
+
+    see: https://github.com/stac-extensions/projection/#item-properties-or-asset-fields
 
-    # Proj
+    """
     try:
         epsg = src_dst.crs.to_epsg() or None
     except AttributeError:
         epsg = None
 
-    metadata["proj"] = {
+    meta = {
         "epsg": epsg,
         "geometry": bbox_to_geom(src_dst.bounds),
         "bbox": list(src_dst.bounds),
         "shape": [src_dst.height, src_dst.width],
         "transform": list(src_dst.transform),
     }
-
     # rasterio can't reproduce wkt2
     # ref: https://github.com/mapbox/rasterio/issues/2044
     # if epsg is None:
-    #      metadata["proj"]["wkt2"] = src_dst.crs.to_wkt()
+    #      meta["wkt2"] = src_dst.crs.to_wkt()
+    return meta
 
-    return metadata
+
+def _get_stats(arr: numpy.ma.array, **kwargs: Any) -> Dict:
+    """Calculate array statistics."""
+    sample, edges = numpy.histogram(arr[~arr.mask])
+    return {
+        "statistics": {
+            "mean": arr.mean().item(),
+            "minimum": arr.min().item(),
+            "maximum": arr.max().item(),
+            "stdev": arr.std().item(),
+            "valid_percent": numpy.count_nonzero(~arr.mask)
+            / float(arr.data.size)
+            * 100,
+        },
+        "histogram": {
+            "count": len(edges),
+            "min": edges.min(),
+            "max": edges.max(),
+            "buckets": sample.tolist(),
+        },
+    }
+
+
+def get_raster_info(
+    src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT, MemoryFile],
+    max_size: int = 1024,
+) -> List[Dict]:
+    """Get raster metadata.
+
+    see: https://github.com/stac-extensions/raster#raster-band-object
+
+    """
+    height = src_dst.height
+    width = src_dst.width
+    if max_size:
+        if max(width, height) > max_size:
+            ratio = height / width
+            if ratio > 1:
+                height = max_size
+                width = math.ceil(height / ratio)
+            else:
+                width = max_size
+                height = math.ceil(width * ratio)
+
+    meta: List[Dict] = []
+
+    area_or_point = src_dst.tags()["AREA_OR_POINT"].lower()
+
+    # Missing `bits_per_sample` and `spatial_resolution`
+    for band in src_dst.indexes:
+        value = {
+            "sampling": area_or_point,
+            "data_type": src_dst.dtypes[band - 1],
+            "scale": src_dst.scales[band - 1],
+            "offset": src_dst.offsets[band - 1],
+        }
+        if src_dst.nodata is not None:
+            value["nodata"] = src_dst.nodata
```

Comment:
Validate that `NaN` is a valid value in the STAC extension

Reply:
nodata is optional in the extension, should we only include `nodata` if the image has a nodata value?

Reply:
yep, right now we add it only if it's not None.

my question if more about is `NaN` (a valid nodata value in GDAL) is valid for the spec.

Reply:
The definition of JSON schema does not allow NaN values. I would suggest to extend the @geospatial-jeff comment.
- If nodata is not defined, the raster has no nodata value
- If nodata is `null` then the nodata value is NaN
WDYT?

If you agree, we could modify the spec to describe this interpretation.

Reply:
well I think you can define NaN in JSON so why not use it?

Reply:
I am not a JSON schema expert so I may miss something. I just found in the spec https://datatracker.ietf.org/doc/html/rfc4627#section-2.4 
`Numeric values that cannot be represented as sequences of digits (such as Infinity and NaN) are not permitted.`

and ECMA-262 (section 24.5.2, JSON.stringify, NOTE 4, page 683 of the ECMA-262 pdf at last edit):

`Finite numbers are stringified as if by calling ToString(number). NaN and Infinity regardless of sign are represented as the String null.`

Reply:
excellent, so if NaN I'll set the value to Null. (that's said I don't think it will be reversible, meaning that when reading a item, pystac or other lib will translate null to NaN)

Reply:
🤔, I'm still not sure about this in fact, the test shows that having nodata set to `NaN` pass the validation script. 

I'm a bit worry about setting it to `null` when we have either `NaN` or `inf` (not sure if inf is a valid nodata in gdal anyway), which will then be non reversible. 🤷‍♂️

Reply:
any feedback @emmanuelmathot @geospatial-jeff @kylebarron

Reply:
Perhaps the raster extension should modify the type of `nodata` to either be str or number, and specify that the only valid strs are "NaN", "inf" or "-inf" (if that's the correct enumeration). I believe that will be validate-able in the JSON schema validation, and should solve this issue. Makes parsing the JSON a bit more complicated, but...it is a bit more complicated!

Reply:
https://github.com/stac-extensions/raster/issues/18

Reply:
alright I reverted to set `nodata=NaN`. 

Notes:
- the validation (using jsonschema) pass which is kinda weird because NaN is not a number 🤷‍♂️ 
- rasterio doesn't really support `inf` (or at least it can't create file with nodata = inf, https://github.com/mapbox/rasterio/issues/2227)

Reply:
@vincentsarago why would we want to promote the use of inf/-inf as a nodata value?

Reply:
@sgillies because it's a valid nodata value in GDAL?

Reply:
@sgillies, to be honest I can't remind if I ever user inf/-inf nodata value, but because it seems to be a valid value in GDAL 
https://github.com/OSGeo/gdal/blob/fec15b146f8a750c23c5e765cac12ed5fc9c2b85/autotest/gcore/gdal_stats.py#L266
https://github.com/OSGeo/gdal/blob/fec15b146f8a750c23c5e765cac12ed5fc9c2b85/gdal/gcore/gdalnodatamaskband.cpp#L138-L142
to me it seems the `specification` we are working on should make sure to allow it. The fact I opened an issue on rasterio is just because I found out that I couldn't create a file with `nodata=inf`

Reply:
I was just checking. Often we can improve interoperability in our specs by reducing the number of options.